hikari/impl/event_manager_base.py

```diff
@@ -50,46 +56,130 @@
+        typing.Coroutine[typing.Any, typing.Any, None],
     ]
 
 
 def _default_predicate(_: event_manager.EventT_inv) -> bool:
     return True
 
 
+_CACHE_RESOURCE_ATTRIBUTE = "__CACHE_RESOURCE__"
+_EVENT_TYPES_ATTRIBUTE = "__EVENT_TYPES__"
+
+
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_resource: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
+    """Add metadata to a listener method to indicate when it should be unmarshalled and dispatched."""
+    if isinstance(event_types, typing.Sequence):
+        # dict.fromkeys is used to remove any duplicate entries here
+        event_types = tuple(dict.fromkeys(itertools.chain.from_iterable(e.dispatches() for e in event_types)))
+
+    else:
+        event_types = event_types.dispatches()
+
+    def decorator(method: UnboundMethodT[EventManagerBaseT], /) -> UnboundMethodT[EventManagerBaseT]:
+        setattr(method, _CACHE_RESOURCE_ATTRIBUTE, cache_resource)
+        setattr(method, _EVENT_TYPES_ATTRIBUTE, event_types)
+        return method
+
+    return decorator
+
+
+@attr.frozen()
+class _Consumer:
+    callback: ConsumerT
+    cache: undefined.UndefinedOr[config.CacheComponents]
+    event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]
+
+
 class EventManagerBase(event_manager.EventManager):
     """Provides functionality to consume and dispatch events.
 
     Specific event handlers should be in functions named `on_xxx` where `xxx`
     is the raw event name being dispatched in lower-case.
     """
 
-    __slots__: typing.Sequence[str] = ("_app", "_listeners", "_consumers", "_waiters")
+    __slots__: typing.Sequence[str] = ("_app", "_dispatches_for_cache", "_listeners", "_consumers", "_waiters")
 
     def __init__(self, app: traits.BotAware) -> None:
         self._app = app
-        self._consumers: typing.Dict[str, ConsumerT] = {}
+        self._dispatches_for_cache: typing.Dict[_Consumer, bool] = {}
+        self._consumers: typing.Dict[str, _Consumer] = {}
         self._listeners: ListenerMapT[base_events.Event] = {}
         self._waiters: WaiterMapT[base_events.Event] = {}
 
         for name, member in inspect.getmembers(self):
             if name.startswith("on_"):
-                self._consumers[name[3:]] = member
+                member = typing.cast("MethodT", member)
+                cache_resource = getattr(member, _CACHE_RESOURCE_ATTRIBUTE, undefined.UNDEFINED)
+                event_types = getattr(member, _EVENT_TYPES_ATTRIBUTE, undefined.UNDEFINED)
+                cache_resource = typing.cast("undefined.UndefinedOr[config.CacheComponents]", cache_resource)
+                event_types = typing.cast(
+                    "undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]]", event_types
+                )
```

Comment:
are these `typing.cast` necessary or can we just type them as being that?
```suggestion
                member = typing.cast("MethodT", member)
                cache_resource: undefined.UndefinedOr[config.CacheComponents] = getattr(member, _CACHE_RESOURCE_ATTRIBUTE, undefined.UNDEFINED)
                event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]] = getattr(member, _EVENT_TYPES_ATTRIBUTE, undefined.UNDEFINED)
```

Reply:
Added